File: Problems/0003. Longest Substring Without Repeating Characters/Solution.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int len = s.size();
        if (len == 0) return 0;

        vector<int> map(256, -1); // The length is set to 256 so that the index corresponds to the ASCII value.
        int left = 0, right = 0;
        int maxLen = INT_MIN;
        
        while (right < len) {
            auto c = s[right];
            if (map[c] != -1) left = max(map[c] + 1, left);

            map[c] = right;
            maxLen = max(maxLen, right - left + 1);
            right++;
        }

        return maxLen;
    }
};

File: Problems/0003. Longest Substring Without Repeating Characters/Solution.rs
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let len = s.len();
        if len == 0 {
            return 0;
        }

        let mut map = vec![-1; 256];
        let (mut left, mut right, mut max_len) = (0, 0, 0);
        let s_chars: Vec<char> = s.chars().collect();

        while right < len {
            let c = s_chars[right] as usize;
            if map[c] != -1 {
                left = left.max(map[c] + 1);
            }

            map[c] = right as i32;
            max_len = max_len.max(right as i32 - left + 1);
            right += 1;
        }

        return max_len;
    }
}